tests/unit/sharded_test.cc

```diff
@@ -201,3 +201,32 @@ SEASTAR_THREAD_TEST_CASE(invoke_on_all_sharded_arg) {
     srv.stop().get();
     arg.stop().get();
 }
+
+SEASTAR_THREAD_TEST_CASE(invoke_on_modifiers) {
+    class checker {
+    public:
+        future<> fn(int a) {
```

Comment:
also, might want to rename `fn` to a name which can be distinguished from its non-const cousin. for instance, "access()" versus "mutate()".

and,  this function is never used in this test. is this expected?

when the compiler looks up for a function based on its arguments,  the function which is resolved to, is not dependent to if the enclosing lambda has the `mutable` specifier. it depends on, well, the arguments' types. 

in this case, if the `this` argument which is implicitly passed to the argument is `const`, the const version is used. see https://godbolt.org/z/M77v3Yzbc.

so, i think it'd be better to use different names for `const` function and for non-const function when testing the mutable lambda and non-mutable lambda.

Reply:
> nit, IMHO, the argument is not quite relevant to the problem we are trying to resolve here.

I need to have calling lambda with capture (below), so I capture `int a`. In order not to have unused capture, I pass it here, so the argument is indeed unused. I can plug it the other way, if you prefer

Reply:
> and, this function is never used in this test. is this expected?

The non-const fn is in use. The `const` one isn't, but it's left "for future", see #2278. I can remove it for now

Reply:
> I need to have calling lambda with capture (below)

this is why i am confused. what makes it important to capture something in this context? i assume you want to test something. but what is it?

Reply:
> > and, this function is never used in this test. is this expected?
> 
> The non-const fn is in use. The `const` one isn't, but it's left "for future", see #2278. I can remove it for now

could you please re-read my comment? and probably the linked godbolt example, if you haven't. i commented out the non-const fn, and the sample compiles just fine.

Reply:
> what makes it important to capture something in this context? 

Because lambda without capture, even being mutable, compiles just fine without the fix. I _do_ need non-empty capture here. And if I use `[a]` capture, but don't do anything with `a` inside lambda, compiler warns me about unused lambda capture. So I pass it as argument to `foo::fn(int)`

Reply:
> could you please re-read my comment? and probably the linked godbolt example, if you haven't. i commented out the non-const fn, and the sample compiles just fine.

I don't see how it's related. My code is

```
   srv.invoke_on_all([a] (checker& s) { return s.fn(a); });
```

`srv` is non-const, `checker& s` argument is non-const either, `s.fn(a)` calls for non-const `fn()` (if I remove const overload, it compiles just fine)

Reply:
Maybe this would explain my intent better. I don't try to test if the lambda calls const fn or non const fn depending on whether it's mutable or not. My point is that invoke_on_other() doesn't work with mutable lambdas at all, it just fails to compile like this

```
/home/xemul/src/seastar/include/seastar/core/future.hh:2074:11: error: no type named ‘type’ in ‘struct std::invoke_result<const invoke_on_modifiers::do_run_test_case() const::<lambda(invoke_on_modifiers::do_run_test_case() const::checker&)>&, invoke_on_modifiers::do_run_test_case() const::checker&>’
 2074 |     using futurator = futurize<std::invoke_result_t<Func, Args&&...>>;
      |           ^~~~~~~~~
```

Reply:
I updated the PR with less code in the test. Please, check, hope it demonstrates the problem in a cleaner way

Reply:
> > what makes it important to capture something in this context?
> 
> Because lambda without capture, even being mutable, compiles just fine without the fix. I _do_ need non-empty capture here. And if I use `[a]` capture, but don't do anything with `a` inside lambda, compiler warns me about unused lambda capture. So I pass it as argument to `foo::fn(int)`

thanks. TIL.